oak_channel/src/lib.rs

```diff
@@ -78,7 +78,14 @@ impl InvocationChannel {
 
     pub fn read_message<M: message::Message>(&mut self) -> anyhow::Result<M> {
         let mut encoded_message: Vec<u8> = Vec::new();
-        let mut first_frame = self.inner.read_frame().context("couldn't read frame")?;
+        let first_frame = self
+            .inner
+            .read_frame(|new_len| {
+                encoded_message.resize(new_len, 0);
+                &mut encoded_message[..]
```

Comment:
This section and the implementation of read_frame could use some more documentation to explain the optimization being performed here. What you wrote in the commit message is a good starting point, I would copy it around here.